File: TransactionRegister/Model/Client/PaymentType.m
//
//  PaymentType.m
//  TransactionRegister
//
//  Created by Eric Romrell on 11/1/16.
//  Copyright © 2016 Eric Romrell. All rights reserved.
//

#import "PaymentType.h"

@implementation PaymentType

+(NSArray<PaymentType *> *)paymentTypes {
	return @[
			 [PaymentType typeWithRealType:@"CREDIT"],
			 [PaymentType typeWithRealType:@"DEBIT"],
			 [PaymentType typeWithRealType:@"SAVINGS"],
			 [PaymentType typeWithRealType:@"PERMANENT_SAVINGS"]
			 ];
}

+(PaymentType *)typeWithRealType:(NSString *)realType {
	PaymentType *me = [[self alloc] init];
	me.realType = [realType uppercaseString];
	me.prettyType = [[me.realType stringByReplacingOccurrencesOfString:@"_" withString:@" "] capitalizedString];
	return me;
}

+(PaymentType *)typeFromIndex:(int)index {
	return [PaymentType paymentTypes][index];
}

-(int)orderIndex {
	return (int) [[PaymentType paymentTypes] indexOfObject:self];
}

-(BOOL)isEqual:(id)object {
	return [object isKindOfClass:[PaymentType class]] && [((PaymentType *) object).realType isEqualToString:self.realType];
}

-(NSComparisonResult)compare:(id)object {
	if ([object isKindOfClass:[PaymentType class]]) {
		PaymentType *type = object;
		if (self.orderIndex > type.orderIndex) {
			return NSOrderedDescending;
		} else if (self.orderIndex < type.orderIndex) {
			return NSOrderedAscending;
		} else {
			return NSOrderedSame;
		}
	}
	return NSOrderedSame;
}

@end

File: TransactionRegister/Model/Client/PaymentType.swift
//
//  PaymentType.swift
//  TransactionRegister
//
//  Created by Eric Romrell on 1/19/17.
//  Copyright © 2017 Eric Romrell. All rights reserved.
//

import Foundation

class PaymentType : NSObject {
	private static let PAYMENT_TYPES = [
		PaymentType(realType: "CREDIT"),
		PaymentType(realType: "DEBIT"),
		PaymentType(realType: "SAVINGS"),
		PaymentType(realType: "PERMANENT_SAVINGS")
	];
	var realType : String
	var prettyType : String
	
	init(realType : String) {
		self.realType = realType
		self.prettyType = realType.replacingOccurrences(of: "_", with: " ").capitalized
	}
	
	static func paymentType(index : Int) -> PaymentType {
		return PAYMENT_TYPES[index]
	}
	
	func orderIndex() -> Int {
		let index = PaymentType.PAYMENT_TYPES.index { (paymentType) -> Bool in
			return paymentType.realType == self.realType
		}
		guard let temp = index else {
			print("Invalid payment type: ", self.realType)
			return -1
		}

		return temp;
	}
}
